src/remoteCollection/handleCollectResponse.js

```diff
@@ -62,21 +80,66 @@ function handleCollectResponse(collectResponse) {
   return collectResponse.then((collectRes) => {
     try {
       validateCollectResponse(collectRes);
+
+      /*
+       * If the transform is a string, then we use that function for all
+       * status codes.
+       */
       const tr = collectRes.generatorTemplate.transform;
-      const t = Array.isArray(tr) ? tr.join('\n') : tr;
-      const transformedSamples = evalUtils.safeTransform(t, collectRes);
-      logger.info(`{
-        generator: ${collectRes.name},
-        numSamples: ${transformedSamples.length},
-      }`);
-      enqueue(transformedSamples);
+      if (typeof tr === 'string') { // match all status codes
+        const samplesToEnqueue = evalUtils.safeTransform(tr, collectRes);
+        logger.info({
+          generator: collectRes.name,
+          url: collectRes.url,
+          numSamples: samplesToEnqueue.length,
+        });
+        return enqueue(samplesToEnqueue);
+      }
+
+      /*
+       * The transform is *not* a string, so check for a status code regex
+       * match which maps to a transform for error samples. Use the first one
+       * to match.
+       */
+      Object.keys(tr).forEach((statusMatcher) => {
+        const re = new RegExp(statusMatcher);
+        if (re.test(res.statusCode)) {
+          const t = tr[statusMatcher];
```

Comment:
I do not understand why are we doing this instead of tr.join('\n').

Reply:
transform will no accept an array of strings, which was a hack anyway

if transform is a string, we just execute it in a vm sandbox but when it's an object, we check the key for a regexp match to response's http status code and execute the associated value when the status matches